File: pkg/engine/evaluator.go
package engine

import (
	"github.com/notnil/chess"
)

// Evaluator interface for position evaluation
type Evaluator interface {
	Evaluate(game *chess.Game) float64
}

// MaterialEvaluator evaluates positions based on material balance
type MaterialEvaluator struct {
	pieceValues map[chess.PieceType]float64
}

// NewMaterialEvaluator creates a new material evaluator
func NewMaterialEvaluator() *MaterialEvaluator {
	return &MaterialEvaluator{
		pieceValues: map[chess.PieceType]float64{
			chess.Pawn:   1.0,
			chess.Knight: 3.0,
			chess.Bishop: 3.0,
			chess.Rook:   5.0,
			chess.Queen:  9.0,
			chess.King:   0.0, // King has no material value
		},
	}
}

// Evaluate returns the material evaluation of the position
func (e *MaterialEvaluator) Evaluate(game *chess.Game) float64 {
	// Check for game over states
	outcome := game.Outcome()
	if outcome == chess.WhiteWon {
		return 1000.0
	} else if outcome == chess.BlackWon {
		return -1000.0
	} else if outcome == chess.Draw {
		return 0.0
	}

	board := game.Position().Board()
	score := 0.0

	for sq := chess.A1; sq <= chess.H8; sq++ {
		piece := board.Piece(sq)
		if piece == chess.NoPiece {
			continue
		}

		value := e.pieceValues[piece.Type()]
		if piece.Color() == chess.White {
			score += value
		} else {
			score -= value
		}
	}

	// Return score from the perspective of the side to move
	if game.Position().Turn() == chess.Black {
		score = -score
	}

	return score
}

// PositionalEvaluator adds positional factors to material evaluation
type PositionalEvaluator struct {
	material *MaterialEvaluator
	pst      PieceSquareTables
}

// NewPositionalEvaluator creates a new positional evaluator
func NewPositionalEvaluator() *PositionalEvaluator {
	return &PositionalEvaluator{
		material: NewMaterialEvaluator(),
		pst:      NewPieceSquareTables(),
	}
}

// Evaluate returns the positional evaluation
func (e *PositionalEvaluator) Evaluate(game *chess.Game) float64 {
	// Start with material evaluation
	score := e.material.Evaluate(game)
	
	// Don't add positional factors for game over positions
	if score >= 900 || score <= -900 {
		return score
	}

	board := game.Position().Board()
	positionalScore := 0.0

	// Add piece-square table values
	for sq := chess.A1; sq <= chess.H8; sq++ {
		piece := board.Piece(sq)
		if piece == chess.NoPiece {
			continue
		}

		psValue := e.pst.GetValue(piece, sq)
		if piece.Color() == chess.White {
			positionalScore += psValue
		} else {
			positionalScore -= psValue
		}
	}

	// Add mobility bonus (number of legal moves)
	mobilityBonus := float64(len(game.ValidMoves())) * 0.1

	// Combine scores
	totalScore := score + positionalScore/100.0 + mobilityBonus

	// Return from perspective of side to move
	if game.Position().Turn() == chess.Black {
		totalScore = -totalScore
	}

	return totalScore
}

File: pkg/engine/evaluator.py
"""
Chess position evaluators.

This module provides different evaluation functions for chess positions,
including material and positional evaluators.
"""

import chess
from abc import ABC, abstractmethod
from typing import Dict
import sys
sys.path.append('/workspace')
from pkg.chess.game import Game
from .piece_square_tables import PieceSquareTables


class Evaluator(ABC):
    """Abstract base class for position evaluation."""
    
    @abstractmethod
    def evaluate(self, game: Game) -> float:
        """
        Evaluate the position.
        
        Args:
            game: Game object
            
        Returns:
            Evaluation score from the perspective of the side to move
        """
        pass


class MaterialEvaluator(Evaluator):
    """Evaluates positions based on material balance."""
    
    def __init__(self):
        """Initialize with standard piece values."""
        self.piece_values: Dict[chess.PieceType, float] = {
            chess.PAWN: 1.0,
            chess.KNIGHT: 3.0,
            chess.BISHOP: 3.0,
            chess.ROOK: 5.0,
            chess.QUEEN: 9.0,
            chess.KING: 0.0,  # King has no material value
        }
    
    def evaluate(self, game: Game) -> float:
        """
        Return the material evaluation of the position.
        
        Args:
            game: Game object
            
        Returns:
            Material evaluation from perspective of side to move
        """
        # Check for game over states
        outcome = game.outcome()
        if outcome == 'white_won':
            return 1000.0
        elif outcome == 'black_won':
            return -1000.0
        elif outcome == 'draw':
            return 0.0
        
        board = game.position
        score = 0.0
        
        # Calculate material balance
        for square in chess.SQUARES:
            piece = board.piece_at(square)
            if piece is None:
                continue
            
            value = self.piece_values[piece.piece_type]
            if piece.color == chess.WHITE:
                score += value
            else:
                score -= value
        
        # Return score from the perspective of the side to move
        if board.turn == chess.BLACK:
            score = -score
        
        return score


class PositionalEvaluator(Evaluator):
    """Adds positional factors to material evaluation."""
    
    def __init__(self):
        """Initialize with material evaluator and piece square tables."""
        self.material = MaterialEvaluator()
        self.pst = PieceSquareTables()
    
    def evaluate(self, game: Game) -> float:
        """
        Return the positional evaluation.
        
        Args:
            game: Game object
            
        Returns:
            Positional evaluation from perspective of side to move
        """
        # Start with material evaluation
        score = self.material.evaluate(game)
        
        # Don't add positional factors for game over positions
        if score >= 900 or score <= -900:
            return score
        
        board = game.position
        positional_score = 0.0
        
        # Add piece-square table values
        for square in chess.SQUARES:
            piece = board.piece_at(square)
            if piece is None:
                continue
            
            ps_value = self.pst.get_value(piece, square)
            if piece.color == chess.WHITE:
                positional_score += ps_value
            else:
                positional_score -= ps_value
        
        # Add mobility bonus (number of legal moves)
        mobility_bonus = len(list(board.legal_moves)) * 0.1
        
        # Combine scores
        total_score = score + positional_score / 100.0 + mobility_bonus
        
        # Return from perspective of side to move
        if board.turn == chess.BLACK:
            total_score = -total_score
        
        return total_score